common/cryptfs.c

```diff
@@ -194,6 +281,114 @@ load_crypto_mapping_table(const char *real_blk_name, const char *master_key_asci
+//			00000000-0000-0000-0000-000000000000-etc (252:1)
+//			 └─00000000-0000-0000-0000-000000000000-etc-integrity (252:0)
+//			    └─ (7:3)
+
+//
+//	And:
+//      user@host:~/$ sudo dmsetup table
+//            dmsetup table
+//				0[...]0-etc-integrity: 0 1677721 integrity 7:3 0 32 J 5 journal_sectors:16368 interleave_sectors:32768 buffer_sectors:128 journal_watermark:50 commit_time:10000
+//				0[...]0-etc: 0 1677721 crypt capi:aegis128-random 00000000000000000000000000000000 0 252:0 0 1 integrity:32:aead
+//
+// [1] https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/dm-integrity.html
+// [2] https://wiki.gentoo.org/wiki/Device-mapper#Integrity
+static int
+create_integrity_blk_dev(const char *real_blk_name, const char *name, const unsigned long fs_size)
+{
+	int fd_mapper;
+	int ioctl_ret;
+	int load_count = -1;
+	char create_buffer[DM_INTEGRITY_BUF_SIZE];
+	struct dm_ioctl *create_io;
+	int create_counter;
+
+	//open device mapper
+	if ((fd_mapper = open(DEV_MAPPER, O_RDWR)) < 0) {
+		ERROR("Cannot open device-mapper");
+	}
+
+	// Create blk device
+	//initialize create_io struct
+	create_io = (struct dm_ioctl *)create_buffer;
+	ioctl_init(create_io, DM_INTEGRITY_BUF_SIZE, name, 0);
+
+	for (create_counter = 0; create_counter < TABLE_LOAD_RETRIES; create_counter++) {
+		ioctl_ret = dm_ioctl(fd_mapper, DM_DEV_CREATE, create_io);
+		if (ioctl_ret != 0) {
+			ERROR("Could not create block device: ioctl(DM_DEV_CREATE: '%lu'|int: %d ) ret: %d, errno: %d",
+			      DM_DEV_CREATE, (int)DM_DEV_CREATE, ioctl_ret, errno);
+		} else {
+			DEBUG("Creating block device worked!");
+			break;
+		}
+
+		usleep(500000);
+	}
+
+	if (create_counter >= TABLE_LOAD_RETRIES) {
+		ERROR("Failed to create block device after %d tries", create_counter);
+		goto errout;
+	}
+
+	if (close(fd_mapper) < 0) {
+		ERROR("Cannot close device-mapper");
+	}
+
+	// Format superblock of the device as needed by dmintegrity
+	DEBUG("Formating the Device");
+	// Create 4kb NULL buffer
+	char null_array[4096];
+	memset(&null_array[0], 0, 4096);
+
+	// Format superblock
+	int ret_val = file_write(real_blk_name, &null_array[0], 4096);
+	if (ret_val < 0) {
+		ERROR_ERRNO("Could not write to file! Superblock not clear. Aborting");
+		return -1;
+	}
+
+	//Load Integrity map table
+	DEBUG("Loading Integrity mapping table");
+
+	load_count = load_integrity_mapping_table(real_blk_name, name, fs_size);
+	if (load_count < 0) {
+		ERROR_ERRNO("Error while loading mapping table");
+		goto errout;
+	} else {
+		INFO("Loading integrity map took %d tries", load_count);
+	}
+
+	// Resume this device to activate it
+	DEBUG("Resuming the blk device");
+	ioctl_init(create_io, DM_INTEGRITY_BUF_SIZE, name, 0);
+
+	ioctl_ret = dm_ioctl(fd_mapper, DM_DEV_SUSPEND, create_io);
+	if (ioctl_ret != 0) {
+		ERROR_ERRNO("Cannot resume the dm-integrity device (ioctl ret: %d, errno:%d)",
+			    ioctl_ret, errno);
+		goto errout;
+	}
+
+	//if the program ends here: everything worked
```

Comment:
if the proram would end here we would have a problem, the function ends here.
just remove the comment

Reply:
Done